File: vrp_example.R
# VRP Heuristics Comparison and Hybrid Approach
#
# This script demonstrates the use of three VRP heuristics:
# 1. Nearest Neighbor Heuristic (constructive)
# 2. Clarke-Wright Savings Algorithm (constructive)
# 3. 2-opt Local Search (improvement)

# Source the individual heuristic files
source("nearest_neighbor_heuristic.R")
source("savings_algorithm.R")
source("two_opt_improvement.R")

# Function to print solution summary
print_solution_summary <- function(name, solution, solve_time) {
  cat(sprintf("\n=== %s ===\n", name))
  cat(sprintf("Total Distance: %.2f\n", solution$total_distance))
  cat(sprintf("Vehicles Used: %d\n", solution$num_vehicles_used))
  cat(sprintf("Feasible: %s\n", solution$feasible))
  cat(sprintf("Solve Time: %.4f seconds\n", solve_time))
  
  if (length(solution$unvisited_customers) > 0) {
    unvisited_str <- paste(solution$unvisited_customers, collapse = ", ")
    cat(sprintf("Unvisited customers: %s\n", unvisited_str))
  }
}

# Main comparison function
compare_vrp_heuristics <- function() {
  cat("Vehicle Routing Problem - Heuristics Comparison\n")
  cat("==================================================\n")
  
  # Problem instance
  depot <- c(0, 0)
  customers <- matrix(c(
    4, 4,
    6, 2,
    8, 6,
    2, 8,
    10, 4,
    12, 2,
    14, 6,
    3, 1,
    9, 9,
    11, 8,
    1, 5,
    13, 3
  ), ncol = 2, byrow = TRUE)
  
  demands <- c(3, 5, 2, 4, 6, 3, 4, 2, 5, 3, 4, 2)
  vehicle_capacity <- 15
  num_vehicles <- 4
  
  cat(sprintf("Number of customers: %d\n", nrow(customers)))
  cat(sprintf("Vehicle capacity: %d\n", vehicle_capacity))
  cat(sprintf("Number of vehicles: %d\n", num_vehicles))
  cat(sprintf("Total demand: %d\n", sum(demands)))
  
  # 1. Nearest Neighbor Heuristic
  cat("\n==================================================\n")
  cat("1. NEAREST NEIGHBOR HEURISTIC\n")
  cat("==================================================\n")
  
  start_time <- Sys.time()
  nn_solution <- nearest_neighbor_vrp(depot, customers, demands, vehicle_capacity, num_vehicles)
  nn_time <- as.numeric(difftime(Sys.time(), start_time, units = "secs"))
  
  print_solution_summary("Nearest Neighbor", nn_solution, nn_time)
  
  # Print detailed routes
  distance_matrix_nn <- compute_distance_matrix(depot, customers)
  for (i in seq_along(nn_solution$routes)) {
    route <- nn_solution$routes[[i]]
    route_info <- get_route_info(route, demands, vehicle_capacity, distance_matrix_nn)
    route_str <- paste(route, collapse = " -> ")
    cat(sprintf("  Vehicle %d: Depot -> %s -> Depot\n", i, route_str))
    cat(sprintf("    Distance: %.2f, Load: %d/%d\n", route_info$distance, route_info$load, vehicle_capacity))
  }
  
  # 2. Clarke-Wright Savings Algorithm
  cat("\n==================================================\n")
  cat("2. CLARKE-WRIGHT SAVINGS ALGORITHM\n")
  cat("==================================================\n")
  
  start_time <- Sys.time()
  cw_solution <- savings_algorithm_vrp(depot, customers, demands, vehicle_capacity, num_vehicles)
  cw_time <- as.numeric(difftime(Sys.time(), start_time, units = "secs"))
  
  print_solution_summary("Clarke-Wright Savings", cw_solution, cw_time)
  
  # Print top savings
  cat("\nTop 5 Savings computed:\n")
  top_savings <- head(cw_solution$savings_list, 5)
  for (i in 1:nrow(top_savings)) {
    cat(sprintf("  %d. Customers %d-%d: Savings = %.2f\n", 
                i, top_savings[i, ]$customer_i, top_savings[i, ]$customer_j, top_savings[i, ]$savings_value))
  }
  
  # Print detailed routes
  distance_matrix_cw <- compute_distance_matrix(depot, customers)
  for (i in seq_along(cw_solution$routes)) {
    route <- cw_solution$routes[[i]]
    route_info <- get_route_info(route, demands, vehicle_capacity, distance_matrix_cw)
    route_str <- paste(route, collapse = " -> ")
    cat(sprintf("  Vehicle %d: Depot -> %s -> Depot\n", i, route_str))
    cat(sprintf("    Distance: %.2f, Load: %d/%d\n", route_info$distance, route_info$load, vehicle_capacity))
  }
  
  # 3. Improve solutions with 2-opt
  cat("\n==================================================\n")
  cat("3. 2-OPT IMPROVEMENT\n")
  cat("==================================================\n")
  
  # Improve Nearest Neighbor solution
  cat("\nImproving Nearest Neighbor solution with 2-opt:\n")
  if (length(nn_solution$routes) > 0) {
    start_time <- Sys.time()
    improved_nn <- improve_solution_2opt(nn_solution$routes, depot, customers, demands, vehicle_capacity, max_iterations = 100)
    opt_time <- as.numeric(difftime(Sys.time(), start_time, units = "secs"))
    
    cat(sprintf("Original NN distance: %.2f\n", nn_solution$total_distance))
    cat(sprintf("Improved distance: %.2f\n", improved_nn$total_distance))
    cat(sprintf("Improvement: %.2f\n", nn_solution$total_distance - improved_nn$total_distance))
    cat(sprintf("Improvement time: %.4f seconds\n", opt_time))
  } else {
    cat("No routes to improve in Nearest Neighbor solution\n")
    improved_nn <- NULL
  }
  
  # Improve Clarke-Wright solution
  cat("\nImproving Clarke-Wright solution with 2-opt:\n")
  if (length(cw_solution$routes) > 0) {
    start_time <- Sys.time()
    improved_cw <- improve_solution_2opt(cw_solution$routes, depot, customers, demands, vehicle_capacity, max_iterations = 100)
    opt_time <- as.numeric(difftime(Sys.time(), start_time, units = "secs"))
    
    cat(sprintf("Original CW distance: %.2f\n", cw_solution$total_distance))
    cat(sprintf("Improved distance: %.2f\n", improved_cw$total_distance))
    cat(sprintf("Improvement: %.2f\n", cw_solution$total_distance - improved_cw$total_distance))
    cat(sprintf("Improvement time: %.4f seconds\n", opt_time))
  } else {
    cat("No routes to improve in Clarke-Wright solution\n")
    improved_cw <- NULL
  }
  
  # Summary comparison
  cat("\n==================================================\n")
  cat("FINAL COMPARISON\n")
  cat("==================================================\n")
  
  results <- list()
  if (nn_solution$feasible) {
    final_nn_distance <- ifelse(is.null(improved_nn), nn_solution$total_distance, improved_nn$total_distance)
    results <- append(results, list(list(method = "Nearest Neighbor (+ 2-opt)", 
                                         distance = final_nn_distance, 
                                         vehicles = nn_solution$num_vehicles_used)))
  }
  
  if (cw_solution$feasible) {
    final_cw_distance <- ifelse(is.null(improved_cw), cw_solution$total_distance, improved_cw$total_distance)
    results <- append(results, list(list(method = "Clarke-Wright (+ 2-opt)", 
                                         distance = final_cw_distance, 
                                         vehicles = cw_solution$num_vehicles_used)))
  }
  
  if (length(results) > 0) {
    # Sort results by distance
    distances <- sapply(results, function(x) x$distance)
    sorted_indices <- order(distances)
    
    cat("Ranking by total distance:\n")
    for (i in seq_along(sorted_indices)) {
      idx <- sorted_indices[i]
      result <- results[[idx]]
      cat(sprintf("  %d. %s: %.2f (using %d vehicles)\n", 
                  i, result$method, result$distance, result$vehicles))
    }
  }
  
  return(list(
    nearest_neighbor = nn_solution,
    clarke_wright = cw_solution,
    improved_nn = improved_nn,
    improved_cw = improved_cw
  ))
}

# Function to demonstrate hybrid approach
demonstrate_hybrid_approach <- function() {
  cat("\n======================================================================\n")
  cat("HYBRID APPROACH: BEST CONSTRUCTIVE + 2-OPT + INTER-ROUTE 2-OPT\n")
  cat("======================================================================\n")
  
  # Same problem instance
  depot <- c(0, 0)
  customers <- matrix(c(
    4, 4,
    6, 2,
    8, 6,
    2, 8,
    10, 4,
    12, 2,
    14, 6,
    3, 1,
    9, 9,
    11, 8,
    1, 5,
    13, 3
  ), ncol = 2, byrow = TRUE)
  
  demands <- c(3, 5, 2, 4, 6, 3, 4, 2, 5, 3, 4, 2)
  vehicle_capacity <- 15
  num_vehicles <- 4
  
  # Try both constructive heuristics
  nn_solution <- nearest_neighbor_vrp(depot, customers, demands, vehicle_capacity, num_vehicles)
  cw_solution <- savings_algorithm_vrp(depot, customers, demands, vehicle_capacity, num_vehicles)
  
  # Choose the better initial solution
  if (nn_solution$feasible && cw_solution$feasible) {
    if (nn_solution$total_distance <= cw_solution$total_distance) {
      best_initial <- nn_solution
      best_method <- "Nearest Neighbor"
    } else {
      best_initial <- cw_solution
      best_method <- "Clarke-Wright"
    }
  } else if (nn_solution$feasible) {
    best_initial <- nn_solution
    best_method <- "Nearest Neighbor"
  } else if (cw_solution$feasible) {
    best_initial <- cw_solution
    best_method <- "Clarke-Wright"
  } else {
    cat("No feasible solution found!\n")
    return(NULL)
  }
  
  cat(sprintf("Best initial solution: %s with distance %.2f\n", best_method, best_initial$total_distance))
  
  # Apply improvements
  # Step 1: Intra-route 2-opt
  cat("\nStep 1: Applying intra-route 2-opt...\n")
  step1_solution <- improve_solution_2opt(best_initial$routes, depot, customers, demands, vehicle_capacity, max_iterations = 200)
  
  # Step 2: Inter-route 2-opt
  cat("\nStep 2: Applying inter-route 2-opt...\n")
  final_solution <- inter_route_2opt(step1_solution$routes, depot, customers, demands, vehicle_capacity, max_iterations = 200)
  
  # Summary
  cat(sprintf("\nHYBRID APPROACH RESULTS:\n"))
  cat(sprintf("Initial (%s): %.2f\n", best_method, best_initial$total_distance))
  cat(sprintf("After intra-route 2-opt: %.2f\n", step1_solution$total_distance))
  cat(sprintf("After inter-route 2-opt: %.2f\n", final_solution$total_distance))
  cat(sprintf("Total improvement: %.2f\n", best_initial$total_distance - final_solution$total_distance))
  improvement_pct <- ((best_initial$total_distance - final_solution$total_distance) / best_initial$total_distance) * 100
  cat(sprintf("Improvement percentage: %.1f%%\n", improvement_pct))
  
  cat(sprintf("\nFinal routes:\n"))
  distance_matrix <- compute_distance_matrix(depot, customers)
  for (i in seq_along(final_solution$routes)) {
    route <- final_solution$routes[[i]]
    route_info <- get_route_info(route, demands, vehicle_capacity, distance_matrix)
    route_str <- paste(route, collapse = " -> ")
    cat(sprintf("  Vehicle %d: Depot -> %s -> Depot\n", i, route_str))
    cat(sprintf("    Distance: %.2f, Load: %d/%d (%.1f%%)\n", 
                route_info$distance, route_info$load, vehicle_capacity, 
                route_info$capacity_utilization * 100))
  }
  
  return(final_solution)
}

# Function to print heuristics summary
print_heuristics_summary <- function() {
  cat("\n======================================================================\n")
  cat("HEURISTICS SUMMARY\n")
  cat("======================================================================\n")
  cat("1. Nearest Neighbor: Fast, greedy constructive heuristic\n")
  cat("   - Good for quick solutions\n")
  cat("   - May not find optimal routes\n")
  cat("\n")
  cat("2. Clarke-Wright Savings: Classical VRP heuristic\n")
  cat("   - Considers global savings\n")
  cat("   - Often produces better initial solutions\n")
  cat("\n")
  cat("3. 2-opt Local Search: Improvement heuristic\n")
  cat("   - Refines existing solutions\n")
  cat("   - Can be applied to any initial solution\n")
  cat("   - Both intra-route and inter-route variants\n")
  cat("\n")
  cat("4. Hybrid Approach: Combines the best of all methods\n")
  cat("   - Use best constructive heuristic as starting point\n")
  cat("   - Apply multiple improvement phases\n")
  cat("   - Generally produces the best results\n")
}

# Main execution function
main <- function() {
  # Run comparison of individual heuristics
  results <- compare_vrp_heuristics()
  
  # Demonstrate hybrid approach
  hybrid_result <- demonstrate_hybrid_approach()
  
  # Print summary
  print_heuristics_summary()
  
  return(list(
    comparison_results = results,
    hybrid_result = hybrid_result
  ))
}

# Run the main function if script is executed directly
if (!interactive()) {
  main()
}

File: vrp_example.py
"""
Example demonstrating the use of three VRP heuristics:
1. Nearest Neighbor Heuristic (constructive)
2. Clarke-Wright Savings Algorithm (constructive)
3. 2-opt Local Search (improvement)

This script compares the performance of different approaches and shows
how to combine constructive and improvement heuristics.
"""

import time
from nearest_neighbor_heuristic import NearestNeighborVRP
from savings_algorithm import SavingsAlgorithmVRP
from two_opt_improvement import TwoOptVRP


def print_solution_summary(name: str, solution: dict, solve_time: float):
    """Print a summary of the solution."""
    print(f"\n=== {name} ===")
    print(f"Total Distance: {solution['total_distance']:.2f}")
    print(f"Vehicles Used: {solution['num_vehicles_used']}")
    print(f"Feasible: {solution['feasible']}")
    print(f"Solve Time: {solve_time:.4f} seconds")
    
    if solution['unvisited_customers']:
        print(f"Unvisited customers: {solution['unvisited_customers']}")


def compare_vrp_heuristics():
    """Compare different VRP heuristics on the same problem instance."""
    
    # Problem instance
    depot = (0, 0)
    customers = [
        (4, 4), (6, 2), (8, 6), (2, 8), (10, 4), (12, 2),
        (14, 6), (3, 1), (9, 9), (11, 8), (1, 5), (13, 3)
    ]
    demands = [3, 5, 2, 4, 6, 3, 4, 2, 5, 3, 4, 2]
    vehicle_capacity = 15
    num_vehicles = 4
    
    print("Vehicle Routing Problem - Heuristics Comparison")
    print("=" * 50)
    print(f"Number of customers: {len(customers)}")
    print(f"Vehicle capacity: {vehicle_capacity}")
    print(f"Number of vehicles: {num_vehicles}")
    print(f"Total demand: {sum(demands)}")
    
    # 1. Nearest Neighbor Heuristic
    print("\n" + "="*50)
    print("1. NEAREST NEIGHBOR HEURISTIC")
    print("="*50)
    
    start_time = time.time()
    nn_solver = NearestNeighborVRP(depot, customers, demands, vehicle_capacity, num_vehicles)
    nn_solution = nn_solver.solve()
    nn_time = time.time() - start_time
    
    print_solution_summary("Nearest Neighbor", nn_solution, nn_time)
    
    # Print detailed routes
    for i, route in enumerate(nn_solution['routes']):
        route_info = nn_solver.get_route_info(route)
        print(f"  Vehicle {i+1}: Depot -> {' -> '.join(map(str, route))} -> Depot")
        print(f"    Distance: {route_info['distance']:.2f}, Load: {route_info['load']}/{vehicle_capacity}")
    
    # 2. Clarke-Wright Savings Algorithm
    print("\n" + "="*50)
    print("2. CLARKE-WRIGHT SAVINGS ALGORITHM")
    print("="*50)
    
    start_time = time.time()
    cw_solver = SavingsAlgorithmVRP(depot, customers, demands, vehicle_capacity, num_vehicles)
    cw_solution = cw_solver.solve()
    cw_time = time.time() - start_time
    
    print_solution_summary("Clarke-Wright Savings", cw_solution, cw_time)
    
    # Print top savings
    print(f"\nTop 5 Savings computed:")
    for i, (cust_i, cust_j, savings) in enumerate(cw_solver.savings_list[:5]):
        print(f"  {i+1}. Customers {cust_i}-{cust_j}: Savings = {savings:.2f}")
    
    # Print detailed routes
    for i, route in enumerate(cw_solution['routes']):
        route_info = cw_solver.get_route_info(route)
        print(f"  Vehicle {i+1}: Depot -> {' -> '.join(map(str, route))} -> Depot")
        print(f"    Distance: {route_info['distance']:.2f}, Load: {route_info['load']}/{vehicle_capacity}")
    
    # 3. Improve solutions with 2-opt
    print("\n" + "="*50)
    print("3. 2-OPT IMPROVEMENT")
    print("="*50)
    
    # Improve Nearest Neighbor solution
    print("\nImproving Nearest Neighbor solution with 2-opt:")
    if nn_solution['routes']:
        start_time = time.time()
        opt_solver = TwoOptVRP(depot, customers, demands, vehicle_capacity)
        improved_nn = opt_solver.improve_solution(nn_solution['routes'], max_iterations=100)
        opt_time = time.time() - start_time
        
        print(f"Original NN distance: {nn_solution['total_distance']:.2f}")
        print(f"Improved distance: {improved_nn['total_distance']:.2f}")
        print(f"Improvement: {nn_solution['total_distance'] - improved_nn['total_distance']:.2f}")
        print(f"Improvement time: {opt_time:.4f} seconds")
    else:
        print("No routes to improve in Nearest Neighbor solution")
    
    # Improve Clarke-Wright solution
    print("\nImproving Clarke-Wright solution with 2-opt:")
    if cw_solution['routes']:
        start_time = time.time()
        improved_cw = opt_solver.improve_solution(cw_solution['routes'], max_iterations=100)
        opt_time = time.time() - start_time
        
        print(f"Original CW distance: {cw_solution['total_distance']:.2f}")
        print(f"Improved distance: {improved_cw['total_distance']:.2f}")
        print(f"Improvement: {cw_solution['total_distance'] - improved_cw['total_distance']:.2f}")
        print(f"Improvement time: {opt_time:.4f} seconds")
    else:
        print("No routes to improve in Clarke-Wright solution")
    
    # Summary comparison
    print("\n" + "="*50)
    print("FINAL COMPARISON")
    print("="*50)
    
    results = []
    if nn_solution['feasible']:
        final_nn_distance = improved_nn['total_distance'] if 'improved_nn' in locals() else nn_solution['total_distance']
        results.append(("Nearest Neighbor (+ 2-opt)", final_nn_distance, nn_solution['num_vehicles_used']))
    
    if cw_solution['feasible']:
        final_cw_distance = improved_cw['total_distance'] if 'improved_cw' in locals() else cw_solution['total_distance']
        results.append(("Clarke-Wright (+ 2-opt)", final_cw_distance, cw_solution['num_vehicles_used']))
    
    if results:
        results.sort(key=lambda x: x[1])  # Sort by distance
        print("Ranking by total distance:")
        for i, (method, distance, vehicles) in enumerate(results):
            print(f"  {i+1}. {method}: {distance:.2f} (using {vehicles} vehicles)")
    
    return {
        'nearest_neighbor': nn_solution,
        'clarke_wright': cw_solution,
        'improved_nn': improved_nn if 'improved_nn' in locals() else None,
        'improved_cw': improved_cw if 'improved_cw' in locals() else None
    }


def demonstrate_hybrid_approach():
    """Demonstrate a hybrid approach combining multiple heuristics."""
    print("\n" + "="*70)
    print("HYBRID APPROACH: BEST CONSTRUCTIVE + 2-OPT + INTER-ROUTE 2-OPT")
    print("="*70)
    
    # Same problem instance
    depot = (0, 0)
    customers = [
        (4, 4), (6, 2), (8, 6), (2, 8), (10, 4), (12, 2),
        (14, 6), (3, 1), (9, 9), (11, 8), (1, 5), (13, 3)
    ]
    demands = [3, 5, 2, 4, 6, 3, 4, 2, 5, 3, 4, 2]
    vehicle_capacity = 15
    num_vehicles = 4
    
    # Try both constructive heuristics
    nn_solver = NearestNeighborVRP(depot, customers, demands, vehicle_capacity, num_vehicles)
    nn_solution = nn_solver.solve()
    
    cw_solver = SavingsAlgorithmVRP(depot, customers, demands, vehicle_capacity, num_vehicles)
    cw_solution = cw_solver.solve()
    
    # Choose the better initial solution
    if nn_solution['feasible'] and cw_solution['feasible']:
        if nn_solution['total_distance'] <= cw_solution['total_distance']:
            best_initial = nn_solution
            best_method = "Nearest Neighbor"
        else:
            best_initial = cw_solution
            best_method = "Clarke-Wright"
    elif nn_solution['feasible']:
        best_initial = nn_solution
        best_method = "Nearest Neighbor"
    elif cw_solution['feasible']:
        best_initial = cw_solution
        best_method = "Clarke-Wright"
    else:
        print("No feasible solution found!")
        return
    
    print(f"Best initial solution: {best_method} with distance {best_initial['total_distance']:.2f}")
    
    # Apply improvements
    opt_solver = TwoOptVRP(depot, customers, demands, vehicle_capacity)
    
    # Step 1: Intra-route 2-opt
    print("\nStep 1: Applying intra-route 2-opt...")
    step1_solution = opt_solver.improve_solution(best_initial['routes'], max_iterations=200)
    
    # Step 2: Inter-route 2-opt
    print("\nStep 2: Applying inter-route 2-opt...")
    final_solution = opt_solver.inter_route_2opt(step1_solution['routes'], max_iterations=200)
    
    # Summary
    print(f"\nHYBRID APPROACH RESULTS:")
    print(f"Initial ({best_method}): {best_initial['total_distance']:.2f}")
    print(f"After intra-route 2-opt: {step1_solution['total_distance']:.2f}")
    print(f"After inter-route 2-opt: {final_solution['total_distance']:.2f}")
    print(f"Total improvement: {best_initial['total_distance'] - final_solution['total_distance']:.2f}")
    print(f"Improvement percentage: {((best_initial['total_distance'] - final_solution['total_distance']) / best_initial['total_distance'] * 100):.1f}%")
    
    print(f"\nFinal routes:")
    for i, route in enumerate(final_solution['routes']):
        route_info = opt_solver.get_route_info(route)
        print(f"  Vehicle {i+1}: Depot -> {' -> '.join(map(str, route))} -> Depot")
        print(f"    Distance: {route_info['distance']:.2f}, Load: {route_info['load']}/{vehicle_capacity} ({route_info['capacity_utilization']:.1%})")


if __name__ == "__main__":
    # Run comparison of individual heuristics
    results = compare_vrp_heuristics()
    
    # Demonstrate hybrid approach
    demonstrate_hybrid_approach()
    
    print("\n" + "="*70)
    print("HEURISTICS SUMMARY")
    print("="*70)
    print("1. Nearest Neighbor: Fast, greedy constructive heuristic")
    print("   - Good for quick solutions")
    print("   - May not find optimal routes")
    print()
    print("2. Clarke-Wright Savings: Classical VRP heuristic")
    print("   - Considers global savings")
    print("   - Often produces better initial solutions")
    print()
    print("3. 2-opt Local Search: Improvement heuristic")
    print("   - Refines existing solutions")
    print("   - Can be applied to any initial solution")
    print("   - Both intra-route and inter-route variants")
    print()
    print("4. Hybrid Approach: Combines the best of all methods")
    print("   - Use best constructive heuristic as starting point")
    print("   - Apply multiple improvement phases")
    print("   - Generally produces the best results")